src/vmhistogram.rs

```diff
@@ -0,0 +1,328 @@
+// Originally distributed under the following license:
+// The MIT License (MIT)
+//
+// Copyright (c) 2019 VictoriaMetrics
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+use std::{
+    mem::MaybeUninit,
+    sync::{Arc, LazyLock},
+};
+
+use parking_lot::Mutex;
+
+use crate::{
+    core::{Collector, Desc, Describer, Metric, MetricVec, MetricVecBuilder},
+    proto::{self, LabelPair},
+    value::make_label_pairs,
+    Opts,
+};
+
+const E_10_MIN: i32 = -9;
+const E_10_MAX: i32 = 18;
+const BUCKETS_PER_DECIMAL: usize = 18;
+const DECIMAL_BUCKETS_COUNT: usize = (E_10_MAX - E_10_MIN) as usize;
+const BUCKETS_COUNT: usize = DECIMAL_BUCKETS_COUNT * BUCKETS_PER_DECIMAL;
+static BUCKET_MULTIPLIER: LazyLock<f64> =
+    LazyLock::new(|| 10f64.powf(1.0f64 / BUCKETS_PER_DECIMAL as f64));
+static LOWER_BUCKET_RANGE: LazyLock<String> =
+    LazyLock::new(|| format!("0...{:.3e}", 10f64.powi(E_10_MIN)));
+static UPPER_BUCKET_RANGE: LazyLock<String> =
+    LazyLock::new(|| format!("{:.3e}...+Inf", 10f64.powi(E_10_MAX)));
+static BUCKET_RANGES: LazyLock<[String; BUCKETS_COUNT]> = LazyLock::new(|| {
+    // SAFETY: The `assume_init` is safe because the type we are claiming to
+    // have initialized here is a bunch of `MaybeUninit`s, which do not require
+    // initialization.
+    let mut ranges: [MaybeUninit<String>; BUCKETS_COUNT] =
+        unsafe { MaybeUninit::uninit().assume_init() };
+    let mut v = 10f64.powi(E_10_MIN as i32);
+    let mut start = format!("{:.3e}", v);
+    for elem in &mut ranges {
+        v = v * *BUCKET_MULTIPLIER;
+        let end = format!("{:.3e}", v);
+        elem.write(format!("{}...{}", start, end));
+        start = end;
+    }
+    // SAFETY: We've iterated over the full array and initialized every element with
+    // a `String`.
+    unsafe { std::mem::transmute(ranges) }
+});
+
+#[derive(Default)]
+struct Inner {
```

Comment:
some notes on the data structure --

1. this does take ~4KB of RAM per histogram since it's storing ~27 * 18 `u64`s. I think that's fine for our application, though.
2. since this is using nested arrays rather than an array of pointers (`[decimalBucketscount]*[bucketsPerDecimal]uint64`), the `Option` doesn't buy us much and we should just make it `[[u64; BUCKETS_PER_DECIMAL]; DECIMAL_BUCKETS_COUNT]`. or, if we want to recover some sparsity, we could do `[Option<Box<[u64; BUCKETS_PER_DECIMAL]>>; DECIMAL_BUCKETS_COUNT]` to match the go implementation.
3. if we decide to stick with the dense array, we could make this lockfree by removing the `Mutex` on inner and changing the `u64` to `AtomicU64`.

Reply:
I'll go with the `Box`ed version for now (it's requires no other code change) and brings us down to 240 bytes for an empty histogram. 

The atomic lockfree version is actually a little trickier than that IIUC — the crate's regular histogram implementation does some fancy hot/cold sharding to satisfy the consistency requirements: https://github.com/tikv/rust-prometheus/blob/master/src/histogram.rs#L293